_posts/2022-09-08-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,168 @@
+
+If it does and this early termination happens conditionally within a function, the optimizer could incorrectly remove storage writes before calls to the function.
+
+More concretely, the bug can occur if a contract contains the following pattern:
+
+1. A storage write. Note that the write may occur only conditionally or within a call to a function that may by inlined.
+2. A call to a function that conditionally terminates using inline assembly as described above, but also has a different code path that returns to the caller.
+3. Any continuing control flow path does one of the following:
+    - It overwrites the storage write in (1).
+    - It reverts.
+
+With legacy code generation, all of this, i.e. steps 1-3, need to happen in a single inline assembly block and the function call in (2) needs to be a call to a user-defined assembly function (see below for examples of affected Yul snippets). Also note that during legacy code generation, the Yul optimizer is only run on inline assembly snippets that do not refer to Solidity variables, further reducing the amount of potentially affected cases.
+
+However, when using via-IR code generation the entire contract is first translated to Yul and then optimized as a whole. In this case, 1-3 can happen in Solidity code and only the function called in (2) needs to involve an inline assembly block calling ``return(...)`` or ``stop()`` (the compiler will not generate ``return(...)`` or ``stop()`` instructions that can trigger the bug without the use of inline assembly). Note that due to inlining, this inline assembly block can also merely occur in another nested function call, i.e. any call in (2) that, through any chain of nested calls, can both return to the caller and terminate via assembly is susceptible.
+
+The following is a minimal example that will exhibit the bug if compiled via IR with enabled optimizer:
+
+```solidity
+contract C {
+	uint public x;
+	function f(bool a) public {
+		x = 1; // This write used to be incorrectly removed by the Yul optimizer due to the bug.  
+		g(a);
+		x = 2;
+	}
+	function g(bool a) internal {
+		// The relevant part of this function is that it can both return to the caller and terminate the transaction.
+		// The bug will show its effects in the cases in which the transaction is terminated (i.e. if a is false).
+		// In this case the write x = 1 above will be missing.
+		if (a) return;
+		assembly { return(0,0) }
+	}
+}
+```
+
+Compiling the above via IR with enabled optimizer will result in ``f(false)`` to incorrectly terminate the transaction without modifying ``x``.
+
+For a concrete contract, the bug may still be prevented, in case the optimizer inlines the function before running the problematic optimizer step, e.g. the following contract is almost identical, but unaffected, since ``g()`` will be inlined (in the snippet above the Solidity-level ``return;`` prevents easy inlining):
```

Comment:
```suggestion
For a concrete contract, the bug may still be avoided if the optimizer inlines the function before running the problematic optimizer step, e.g. the following contract is almost identical, but unaffected, since ``g()`` will be inlined (in the snippet above the Solidity-level ``return;`` prevents easy inlining):
```

Reply:
To me "avoiding" sounds more like something somebody does actively... whereas "being prevented" is something that just happens... but also fine with changing...

Reply:
Haha, I had the exact opposite impression. I.e. you actively prevent something but in this case it's not under your control so you may just passively avoid it.

In any case, not very important. Just sounded better to me that way.

Reply:
Yeah, you have a point there - can be read that way as well :-). The real problem is probably phrasing it as passive... let me rephrase it after all - if I say ``the optimizer may still prevent the bug if it inlines``, it's clear. I still prefer ``prevent`` there, not ``avoid`` for some reason, but fine either way :-D.